### ISSUE TITLE
Procedures for quantizing generic architectures

### ISSUE THREAD

**Issue by ad8e** [Cornell-RelaxML/quip-sharp#45]
I have a custom 13B model I'm trying to quantize with QuIP#. In the `quantize_llama` scripts, llama2's architecture is hardcoded in a few places. It seems brittle and difficult for me to try to use my architecture with these, since I won't be able to track upstream changes, and I won't know when I have made all the necessary changes.

Many other quantization libraries have the same issue, but there have been two solutions:

1. Some libraries let us specify the layer names to quantize, such as: https://github.com/neuralmagic/sparseml/tree/main/src/sparseml/transformers/sparsification/obcq#-how-to-adapt-a-recipe-for-a-new-model So when we want to test new architectures, we only need to change the layer names.
2. Gpt-fast automatically deduces the layer configuration using torch._dynamo.export: https://github.com/pytorch-labs/gpt-fast/blob/40ff31458beffa3ddda8c45215240d6ce43a768a/GPTQ.py#L134

If either of these solutions is possible with QuIP#, it would be convenient for experimenting with new transformer architectures.

**Comment by tsengalb99** [comment]
There shouldn’t be too much hardcoded stuff in the quantize_llama scripts. I’m aware the hessian script might be a bit hardcoded since I didn’t get around to updating that, but the general recipe for all these scripts is:

*	Construct the attention mask and position IDs so we can do things layer by layer
*	Identify the relevant linear layers in the models
*	Do stuff

The first step can be done by looking at the forward function in the huggingface implementation of the model and the second step can be done by inspecting the model architecture. QuIP# is an ongoing research project so I haven’t been able to invest much time into making this a fully automated quantization package, but the ideas and algorithms are definitely adaptable to different architectures.

 

From: Kevin Yin ***@***.***> 
Sent: Saturday, February 24, 2024 2:33 PM
To: Cornell-RelaxML/quip-sharp ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [Cornell-RelaxML/quip-sharp] Procedures for quantizing generic architectures (Issue #45)

 

I have a custom 13B model I'm trying to quantize with QuIP#. In the quantize_llama scripts, llama2's architecture is hardcoded in a few places. It seems brittle and difficult for me to try to use my architecture with these, since I won't be able to track upstream changes, and I won't know when I have made all the necessary changes.

Many other quantization libraries have the same issue, but there have been two solutions:

1.	Some libraries let us specify the layer names to quantize, such as: https://github.com/neuralmagic/sparseml/tree/main/src/sparseml/transformers/sparsification/obcq#-how-to-adapt-a-recipe-for-a-new-model So when we want to test new architectures, we only need to change the layer names.
2.	Gpt-fast automatically deduces the layer configuration using torch._dynamo.export: https://github.com/pytorch-labs/gpt-fast/blob/40ff31458beffa3ddda8c45215240d6ce43a768a/GPTQ.py#L134

If either of these solutions is possible with QuIP#, it would be convenient for experimenting with new transformer architectures.
